### ISSUE TITLE
Onready setters

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#325]
**Describe the project you are working on:**
Game with "interactive" nodes, which use tool script with setters and stuff.

**Describe the problem or limitation you are having in your project:**
Real use case here. I have a node called Door. It has two exported properties with setters: `right` and `texture`. When you set `right` to true, the sprite gets flipped, but there's also a collider that changes position to opposite side. `texture` is a... well, sprite texture.

The Door script is a tool script. I recently noticed that I can't just use `_process()` in every tool script that reacts to variable changes, because it kills my GPU due to constant scene redrawing. So I switched to setters and ran into another problem. As you might guess from above description, my tool script affects two underlying nodes by editing their properties. The problem is initialization. When I run the game, the setters are called instantly and error out, because they rely on using `get_node()` before the children are initialized. Right now I'm using a workaround where in the setter I check if the node is inside tree and `return` if not and also inside `_ready()` I call the setters directly _with current value_ to initialize them properly. tbh it feels like a hack and also I will have to do this for every "interactive" object, which is meh.

Related issues:
https://github.com/godotengine/godot/issues/30460
https://github.com/godotengine/godot/issues/28014

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
My idea is being able to defer the call of setters. Consider this syntax:
```
export var right: bool onready setget set_right
```
Notice the `onready` before setget (this might look ugly, but I don't have idea for better syntax rn). It means that when initializing the object, the engine should use `call_deferred` instead of `call` on the setter. This way it will be called after the node is inside tree and have its children properly initialized, so it can modify them.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![image](https://user-images.githubusercontent.com/2223172/71278669-85f8f100-2358-11ea-9f90-7f834e702a7a.png)
maybe

**Describe implementation detail for your proposal (in code), if possible:**
Hopefully it should be possible to call the setter in deferred mode. However this might happen after `_ready()` (unless we use some new logic, but that makes it complicated), so maybe the property could have its value set bypassing the setter and then call the setter on idle frame. Sounds perfectly possible and solves the problem.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It's a change to GDScript, so non-pluginable.

**Is there a reason why this should be core and not an add-on in the asset library?:**
> non-pluginable

**Comment by willnationsdev** [comment]
I have encountered this myself when writing tool scripts. It is definitely frustrating to have to deal with, and I wrote a hack *very* similar to yours.

**Comment by Toshiwoz** [comment]
Hey, I did have also problems with setters getters execution order, and I actually have a doubt.
If you write:
`export onready var right: bool onready setget set_right`
I think it should set the values later than without the `onready` keyword. But I'm not sure about the exact order.

**Comment by Calinou** [comment]
> I recently noticed that I can't just use _process() in every tool script that reacts to variable changes, because it kills my GPU due to constant scene redrawing.

This sounds like a bug to me. Redrawing should only be requested if the property's value actually changes. Can you post the script that triggers constant redrawing?

**Comment by KoBeWi** [comment]
> This sounds like a bug to me. Redrawing should only be requested if the property's value actually changes. Can you post the script that triggers constant redrawing?

```
$Sprite.texture = texture
$Sprite.flip_h = !right
$Solid.position.x = abs($Solid.position.x) * (-1 if right else 1) #Solid is a StaticBody2D
```
And few similar others.

**Comment by pycbouh** [comment]
I'm not sure what's the intended behavior here when working with such property right after it was supposedly set? There are two proposed ways to deal with setters: deferring the call to them internally or bypassing them entirely until the node receives an "idle_frame" signal. The first one means to me that if we try to access the value immediately, it's going to be stale:

```gdscript
some_node.right = true
print(some_node.right) # <-- Value did not visibly change
```

At least it's going to be stale until we enter the tree here. This is completely hidden from the user of the node, because the only indicator here is the new keyword **inside** the node's script. This would be extremely confusing to deal with.

The second one suggests that we store externally provided value without passing it through the setter and possibly sanitizing it? That's even worse, in my opinion. A setter function can do a million things alongside accessing child nodes, thus bypassing it completely is going to put any internal logic in there in danger. And here as well this behavior is not expected by the user of the node.

And the same problem applies to getters in both suggestions: are they bypassed as well until the node is inside the tree? They definitely cannot be deferred... In the end, to me these proposed changes allow for bad code to happen and it puts a lot of faith into the developer who creates these nodes.

We are talking about properties exposed at the top of some scene that abstracts a whole bunch of other nodes. Setters give you an opportunity to take care of these properties in a seamless way and handle various cases the end-user may not be aware about. Such as the node needing to be inside the tree to have the value applied. I don't see what's so hacky or limited about our current capabilities. Frankly, I do this all the time and don't feel like it's a hack:

```gdscript
onready var right : bool = false setget set_right

func _ready() -> void:
  _update_sprite()

func _update_sprite() -> void:
  if (!is_inside_tree()):
    return
  
  $Sprite.flip_h = !right

func set_right(value : bool) -> void:
  right = value
  _update_sprite()
```

I guess it's essentially the same as you've described your approach, but in some different words. Yet using this you can implement your deferred behavior right now, by deferring the call to `_update_sprite`, if that's what you want. This may be wordy, but it's robust. And does not cause any unexpected behavior on exported/public properties.

Please correct me if I've misunderstood what's on the table here.

**Comment by Shadowblitz16** [comment]
I honestly think onready, and _enter_tree and _exit_tree should be made obsolete.
I don't see a point in having the ability to do things before the tree is ready you can literally do everything in ready

Instead make everything run once the tree is completely ready and then add a _deleted hook and deleted signal for when nodes are removed or queue_freed

**Comment by Calinou** [comment]
@Shadowblitz16 `_exit_tree()` is useful to run code when your project exits. For instance, this can be used to save the current game or user configuration automatically:

```gdscript
extends Node

func _exit_tree():
    # Assuming we have a Settings singleton with a "save" method...
    Settings.save()
```

**Comment by Shadowblitz16** [comment]
I mean a `_removed` would be the same thing wouldn't it?

**Comment by red1939** [comment]
Wouldn't it make sense to treat all properties as `onready`? Currently, it seems that the `setget` is ran before `_ready` so accessing any nested nodes will fail (as probably everyone knows). If we were do delay them, no `if`s would be needed.

**Comment by dalexeev** [comment]
@red1939 No, this is an artificial limitation, especially if you remember that there are scripts that are not inherited from `Node`.

**Comment by cgbeutler** [comment]
I wrote this comment for folks that want to fix things the "godot" way. Do not continue reading if that sounds... angering? My intent is not to start a fight, but to help folks learn the engine the way it was built. If you want to fight to change the way the engine works, that's up to you. I am not here for that fight, but education on things as they are now.

When exporting, you get both disk storage AND an editor view. This means that anything with `export` is saved/loaded automagically. In `tool` mode, this also means the value will be copied over if the editor detects a script change and has to rebuild the object. (Without export in tool mode, the variable data can just vanish if you change the script and that variable isn't in the init stuff.)

With that in mind, the feature request seems to signify that you have inverted your data dependency for the tree. You have child nodes requiring a parent's info. Godot was not _really_ built for inverting data dependencies to flow down the tree. The idea was to ensure that any child could always be split out into a little scene of it's own without it breaking.

So, how do we fix the data dependency direction? Depends on who all needs the shared data.

1. If you are just trying to pass up a setter to the parent level, and the child also exports that data, you have actually duplicated the saving and loading of that data on accident. With both having the `export` keyword, you are storing the data to disk twice and it is then loaded twice. If those two exports get out of sync, you could end up with corrupt data. What you likely want is to export the parent's var _without_ storage. That would give it the editor view, but only the child would save it. To do that, you currently have to override `_get_property_list()` and return something like: `[ { name="right", type=TYPE_BOOL, usage=PROPERTY_USAGE_EDITOR} ]`. That will export the variable `right` without the "storage" part. That means the variable's setter will not be called on load, cuz it wasn't saved to the scene to begin with. With this fix in place, only the child will save the data and your data dependency tree is restored! (Note that eventually this will get an attribute so you won't have to go through `_get_property_list`.)

2. If you have a bit of data that is used by several nodes in a scene, you should move that data to a spot of it's own. By moving the data out of the parent and children altogether, all those nodes can all depend on the new spot to provide the data. The intended mechanism for shared data among many nodes is a custom Resource. There's a lot to custom resources that I can't cover here, (pretty sure I could write a book on them.) You have likely already used Resources a lot. All textures are Resources. They can be saved to a scene or to the filesystem for more of a globally-shared-data location. Resources are the intended spot for shared data and even come with a "changed" signal for notifying dependents on when their data changes. Their other strength is that they come with caching, so if you load a texture all over the place, it really only gets loaded once, then cached and shared.

There may be other ways to fix the dependency order, but those are the ones I use the most. Now, let me just reiterate, I am not here to fight this issue. I am just here to share what the engine _does_ offer as solutions. I just want to help folks understand godot a bit better. Ciao!

**Comment by nathanfranke** [comment]
Workaround for Godot 3.x which I've used over 20 times so far:

```py
export var my_var: int setget _set_my_var
func _set_my_var(value: int) -> void:
    my_var = value
    
    if not is_inside_tree():
        yield(self, "ready")
    
    # Custom behaviour
```

Note that this doesn't guarantee the node is ready, only inside the tree, since if you call this from `_enter_tree`, `is_inside_tree` will be `true`.

**Comment by cgbeutler** [comment]
For folks using @nathanfranke's solution, be aware that you can accidentally double-save data if you aren't careful and use it to set fields exported by child nodes.
I recommend actually opening your tscn files when creating exports in order to make sure you are saving what you think you are. Doing so has saved me lots of mistakes. Those files are not as scary as they sound.

When you do `export var foo: int setget set_foo`, the export still creates/uses a getter, even though you didn't put one there. If you open the scene, you'll see it saved the data like so:
```
[node name="Node2D" type="Node2D"]
script = ExtResource( 1 )
foo = 3                            <---- got exported
```

So if you `export` a var, then forward that to a child's `export`ed value, you will see it saved in both places. When the scene is later loaded, it will set it in the child, then again in the parent, lengthening load time and your total file size for no reason.

Sadly, there isn't a quick way to export without `PROPERTY_USAGE_STORAGE` right now. Upvote https://github.com/godotengine/godot-proposals/issues/2539 if you'd like it 😉.
To get "no storage", you'd have to export the long way using `_get_property_list()`.
Though, if you are sharing data across the tree, I'd recommend a Custom Resource instead. They are tricky, but they are the mechanism designed for cross-tree data.

**Comment by KoBeWi** [comment]
A better syntax would be maybe
```GDScript
@onready @export var something
```
and the variable would be initialized like onready variable. Right now when doing this it doesn't affect when the setter is called.

**Comment by LakshayaG73** [comment]
Hi all. 

Any idea how to circumvent this? 

Is the best solution still to use:

`if not is_inside_tree(): await ready`

inside setter functions?

**Comment by KoBeWi** [comment]
Yes.

**Comment by OscarPindaro** [comment]
I'm having the same identical problem.
I have a Sprite node with an exported variable called _collision\_visible_ that handles the visibilitty of a child node that contains various collision shapes. I use a setter that allows to change the value of the visibility without having to interact with the children of the Sprite node. A **tool script** is attached to the sprite node.
`
export(bool) onready var collision_visible : bool = false setget set_collision_visible, get_collision_visible

func set_collision_visible(new_value : bool):
	collision_visible = new_value
        get_node(COLLISIONS_NODE_NAME).set_visible(new_value)
`
As you can see, the default value of _collision\_visible_ is **false**, and if i set it to **true** through the game engine interface, the setter method is called before the child collision node exists, causing an error. This is VERY not intuitive.
I think that it's pretty normal to expect that setters will try to access lower level nodes and their property. 
I don't think that is reasonable to expect that for each setter one should implement a series of checks that control if the child nodes are present in the tree or not.

**Comment by bli00** [comment]
Hey will this be supported in the 4.0 release or nah?

**Comment by YuriSizov** [comment]
No. This is an open proposal, that has not been worked on, let alone merged. And it still has open questions about its implementation.

**Comment by sairam4123** [comment]
> ## 3.x
> Workaround for Godot 3.x which I've used over 20 times so far:
> 
> ```gdscript
> export var my_var: int setget _set_my_var
> func _set_my_var(value: int) -> void:
>     my_var = value
>     
>     if not is_inside_tree():
>         yield(self, "ready")
>     
>     # Custom behaviour
> ```
> 
> Note that this doesn't guarantee the node is ready, only inside the tree, since if you call this from `_enter_tree`, `is_inside_tree` will be `true`.
> 
> **Edit: This version solves the previous problem, albeit a bit more verbose:**
> 
> ```gdscript
> onready var _is_ready := true
> 
> export var my_var: int setget _set_my_var
> func _set_my_var(value: int) -> void:
>     my_var = value
>     
>     if not _is_ready:
>         yield(self, "ready")
>     
>     # Custom behaviour
> ```
> 
> Edit 2:
> 
> ## 4.0
> ```gdscript
> @export var my_var: int:
> 	set(value):
> 	    my_var = value
>     	
>     	if not is_inside_tree():
>         	await ready
>     	
> 	    # Custom behaviour
> ```
> 
> I suggest that if for whatever reason this feature isn't desired, we should at least add an `is_ready()` method. This will fix the issue I noted above and also make the `await` code way more readable.

There is `is_node_ready()` function right?

This would become:
## 4.0
```gdscript
@export var my_var: int:
	set(value):
	    my_var = value
    	
    	    if not is_node_ready():
        	await ready
    	
	    # Custom behaviour
```

**Comment by dalexeev** [comment]
See also:

* #5718
* godotengine/godot#78843

**Comment by ttencate** [comment]
Great trick, @sairam4123! In C#, we can't use `await` in setters, but we can use an extension method on `Node` like this:

```csharp
    public static async void InvokeOnceReady(this Node node, Action action) {
        if (!node.IsNodeReady()) {
            await node.ToSignal(node, Node.SignalName.Ready);
        }
        action.Invoke();
    }
```

Usage:

```csharp
    private Color _color;
    [Export] public Color Color {
        get => _color;
        set {
            _color = color;
            this.InvokeOnceReady(() => GetNode<Polygon2D>("$Polygon2D").Color = Color);
        }
    }
```

**Comment by Shadowblitz16** [comment]
we need onready getters too...
```
@export var value:float:
	get: return $Header/Slider.value
	set(value): $Header/Slider.value = value
```

This throws an error because $Header/Slider is only available after ready.

**Comment by hsandt** [comment]
I have another use case at edit time in my current project, where I'm dynamically creating a node with `@tool` code based on a field via its setter callback:

`created_node.owner = get_tree().edited_scene_root`

But the tree hasn't been assigned by that point and you get an error if you do it too early (read: when you open the scene and the setter callback is immediately called).

I also had to `call_deferred` the creation of the node...

I tried both `if not is_inside_tree(): await ready` and `if not is_node_ready(): await ready`, as well as `if not is_inside_tree(): await tree_entered` (would be more logical and consistent in my case), they all work, however since I was inside a function called from the setter I had to add `await` in front of those function calls as well.

**Comment by Naros** [comment]
Rather than use `get_tree()`, you can use `Engine.get_main_loop()` as a replacement I believe. That will be available to you even at this point.
